### PR TITLE
[MRG] Add silhouette_samples to tslearn.clustering

### LINKED ISSUE #451: Support for  silhouette_samples() functionalities like sklearn.
Is there any support for silhouette_samples() functionalities as sklearn has.

### PR DESCRIPTION
**Pull request by SomSamantray**
## Summary

Time-series clustering users can now get per-sample silhouette coefficients with time-series metrics. Previously only the scalar `silhouette_score` was available; per-sample values (needed for silhouette plots and identifying misassigned or outlier series) required a manual two-step workaround of computing a cross-distance matrix and calling sklearn directly.

`tslearn.clustering.silhouette_samples()` accepts time-series data or a precomputed distance matrix, computes the cross-distance matrix with the requested metric (`dtw`, `softdtw`, `euclidean`, a callable, or `"precomputed"`), and delegates the per-sample computation to sklearn's `silhouette_samples` with `metric="precomputed"` — mirroring the existing `silhouette_score` wrapper. The function deliberately omits `sample_size`/`random_state` (sklearn's `silhouette_samples` has no subsampling), rejecting non-`None` values with a `TypeError`.

## What changed

- New `tslearn.clustering.silhouette_samples()` in `tslearn/clustering/utils.py`, exported from `tslearn.clustering`.
- Per-sample values are `(n_ts,)` and satisfy the documented invariant `mean(silhouette_samples(...)) == silhouette_score(...)` for the same inputs.
- Callable metrics receive `metric_params`/`**kwds` and return backend-normalized values, so a callable behaves like the equivalent string metric under both numpy and pytorch backends.
- Docstring with doctests, API reference entry, CHANGELOG entry, and a silhouette-plot example (`docs/examples/clustering/plot_silhouette.py`).

## Testing

- `test_silhouette_samples()` covers shape, value range, mean-vs-score equality per metric, `TypeError` rejection of `sample_size`/`random_state` across all metric branches, `None` no-op semantics, and `metric_params` forwarding to callable metrics.
- `pytest tests/test_clustering.py --doctest-modules tslearn/clustering/utils.py`: 11 passed.
- Example verified headless (matplotlib Agg): mean silhouette 0.71 on the Trace dataset.

Related: #451

### CODE REVIEW

File: tslearn/clustering/utils.py [hunk 1]
```diff
@@ -215,6 +217,162 @@ def sklearn_metric(x, y):
...
+        sklearn_X = _cdist_dtw(
+            X,
+            n_jobs=n_jobs,
+            verbose=verbose,
+            be=be,
+            **metric_params_
+        )
+    elif metric == "softdtw":
+        X = to_time_series_dataset(X, be=be)
+        sklearn_X = _cdist_soft_dtw_normalized(X, be=be, **metric_params_)
```
Comment (charavelg) [inline tslearn/clustering/utils.py]: `_cdist_soft_dtw_normalized` now have `n_jobs` and `verbose` args (It should also be updated in silhouette_score but that's another matter).
```suggestion
        sklearn_X = _cdist_soft_dtw_normalized(X, be=be, n_jobs=n_jobs, verbose=verbose, **metric_params_)
```

File: tslearn/clustering/utils.py [hunk 2]
```diff
@@ -215,6 +217,162 @@ def sklearn_metric(x, y):
...
+    verbose : int (default: 0)
+        If nonzero, joblib progress messages are printed during the
+        cross-distance matrix computation (only used by the dtw branch).
+
+    **kwds : optional keyword parameters
+        Any further parameters are passed directly to the distance function,
+        just as for the `metric_params` parameter.
+        Note that, unlike :func:`tslearn.clustering.silhouette_score`,
+        ``sample_size`` and ``random_state`` are not supported (they are
+        rejected with a ``TypeError`` if provided).
```
Comment (charavelg) [inline tslearn/clustering/utils.py]: Why is it necessary to reject `sample_size` and `random_state` specifically? Aren't 'kwds' and 'metric_params' arguments redundant? I understand this mimics silhouette_score definition but this seems irrelevant.

File: tslearn/clustering/utils.py [hunk 3]
```diff
@@ -215,6 +217,162 @@ def sklearn_metric(x, y):
...
+    """
+    if kwds.get("sample_size") is not None or \
+            kwds.get("random_state") is not None:
+        raise TypeError(
+            "silhouette_samples does not support the 'sample_size' or "
+            "'random_state' parameters (see sklearn.metrics.silhouette_samples)"
+        )
+    # Explicit None for these keys is a no-op, matching sklearn semantics
+    kwds = {k: v for k, v in kwds.items()
+            if k not in ("sample_size", "random_state")}
```
Comment (charavelg) [inline tslearn/clustering/utils.py]: Same as above

File: docs/residual-review-findings/feat-silhouette-samples.md
Comment (charavelg) [inline docs/residual-review-findings/feat-silhouette-samples.md:1]: Should not be commited

File: docs/examples/clustering/plot_silhouette.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+neighboring clusters, bars near 0 indicate time series close to the decision
+boundary between two clusters, and negative bars indicate time series that
+may have been assigned to the wrong cluster.
+
+.. [1] Peter J. Rousseeuw. "Silhouettes: a Graphical Aid to the
+  Interpretation and Validation of Cluster Analysis". Computational and
+  Applied Mathematics 20: 53-65, 1987.
+"""
+
+# Author: Romain Tavenard
```
Comment (charavelg) [inline docs/examples/clustering/plot_silhouette.py]: Probably not the right author :)

File: docs/examples/clustering/plot_silhouette.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+X_train = X_train[y_train < 4]  # Keep first 3 classes
+numpy.random.shuffle(X_train)
+# Keep only 50 time series
+X_train = TimeSeriesScalerMeanVariance().fit_transform(X_train[:50])
+# Make time series shorter
+X_train = TimeSeriesResampler(sz=40).fit_transform(X_train)
+
+# k-means clustering with DTW
+km = TimeSeriesKMeans(n_clusters=3, metric="dtw", verbose=True,
+                      random_state=seed)
```
Comment (charavelg) [inline docs/examples/clustering/plot_silhouette.py]:
```suggestion
km = TimeSeriesKMeans(n_clusters=3, metric="dtw", random_state=seed)
```
See https://tslearn--703.org.readthedocs.build/en/703/auto_examples/clustering/plot_silhouette.html for generated documentation

**Comment by charavelg** [comment]
Hi @SomSamantray , thanks for the PR, that would be indeed a nice addition to tslearn!!!

Would you mind  checking the review so we can make your PR cross the finish line ? I'm of course available to help you in any way if needed.

**Comment by SomSamantray** [comment]
Hi @charavelg — thanks for the "Cleaner figure" fix, pulled it in and it looks good.

Checked the coverage concern as well: the Codecov patch-coverage gap (2 uncovered lines in tslearn/clustering/utils.py) was closed in e32c94f before this review — current checks show codecov/patch at 100.00% of diff hit and codecov/project at 95.20% (+0.05%), no annotations. All CI is green and the two residual P3 items are already tracked separately as #701 and #702 (non-blocking).

This should be good to merge whenever you get a chance — let me know if you'd like anything else addressed first!

**Comment by charavelg** [comment]
Really sorry I started the review but forgot to submit it ... You should now be able to see it...

File: tslearn/clustering/utils.py (hunk 2 above)
Comment (SomSamantray) [reply tslearn/clustering/utils.py]: Good question — the rejection guard exists because `silhouette_score` accepts `sample_size`/`random_state`, so users naturally try to pass them to the per-sample variant by muscle memory. sklearn's per-sample `silhouette_samples` has no subsampling at all, so a silent no-op would quietly produce results from a subset (or, worse, mislead users into thinking subsampling applied). Rejecting with a `TypeError` makes the unsupported parameter loud.

On `kwds` vs `metric_params`: they are intentionally redundant. `**kwds` mirrors `silhouette_score`'s signature so callers can pass metric kwargs positionally (`silhouette_samples(X, y, metric="dtw", global_constraint="sakoe_chiba")`), while `metric_params` is the canonical dict form. Both feed the same merged dict.

One caveat I'll note: the guard is bypassable by passing `sample_size` inside `metric_params` (it's forwarded to the metric as a kwarg). That's tracked as #701.

File: tslearn/clustering/utils.py (hunk 3 above)
Comment (SomSamantray) [reply tslearn/clustering/utils.py]: Same rationale as the other thread: the `TypeError` guard makes the unsupported `sample_size`/`random_state` (which `silhouette_score` accepts but sklearn's per-sample variant does not support) fail loudly instead of silently no-op'ing. `kwds` and `metric_params` are intentionally redundant convenience channels into the same merged dict.

File: tslearn/clustering/utils.py (hunk 1 above)
Comment (SomSamantray) [reply tslearn/clustering/utils.py]: > `_cdist_soft_dtw_normalized` now have `n_jobs` and `verbose` args

Fixed — the softdtw branch now threads `n_jobs` and `verbose` through to `cdist_soft_dtw_normalized` (the public wrapper, which forwards them), in commits cb36ac1 and 7046fdd. The `verbose` docstring now reflects that it applies to the dtw and softdtw branches.

File: docs/residual-review-findings/feat-silhouette-samples.md
Comment (SomSamantray) [reply docs/residual-review-findings/feat-silhouette-samples.md:1]: > Should not be commited

Fixed — the internal `docs/residual-review-findings/feat-silhouette-samples.md` artifact was removed from the branch in commit 90c6797 and no longer appears in the PR diff.

File: tslearn/clustering/utils.py (hunk 2 above)
Comment (charavelg) [reply tslearn/clustering/utils.py]: On kwds vs metric_params, I do agree with your analysis and we have the same implementation on the tslearn silhouette_score, so let's go that way!

Regarding explicit rejection guard, I don't see this kind of checks within sklearn implementation for example which has the same silhouette_score / silhouette_samples couple. Plus, removing it it would make #701 irrelevant.  

>  a silent no-op would quietly produce results from a subset
Could you explain ? What's the code path that would lead to such a result?

Definitely in favor of removing the explicit guard part so far

Comment (SomSamantray) [reply tslearn/clustering/utils.py]: On the removal: you're right, once I actually tried to construct the silent-no-op scenario I couldn't find one that the guard was protecting against — for dtw/softdtw/callable the value gets forwarded to the underlying cdist call regardless (erroring naturally if unsupported), and for euclidean/precomputed metric_params was never even consulted. So the guard's only power was rejecting *itself*, which isn't a real protection. Removed in `14f930a`, along with the corresponding TypeError tests in test_clustering.py updated to reflect the new (sklearn-matching) behavior.

Comment (charavelg) [reply tslearn/clustering/utils.py]: nice!

File: tslearn/clustering/utils.py (hunk 3 above)
Comment (charavelg) [reply tslearn/clustering/utils.py]: Gone with guard removal

**Comment by charavelg** [comment]
@SomSamantray, thanks for your time and effort making this PR go through!!! Nice addition to tslearn code base!!!
